### ISSUE TITLE
Calculate SHAP values on distributed xgboost

### ISSUE THREAD

**Issue by tien-nguyen** [shap/shap#221]
Hi Scott,

I'm interested in learning more about SHAP and using SHAP values for my project. I read through your example here (https://github.com/slundberg/shap/blob/master/tests/explainers/test_tree.py), especially how you extract shap values at these lines:

`  # train XGBoost model
    X, y = shap.datasets.boston()
    model = xgboost.train({"learning_rate": 0.01}, xgboost.DMatrix(X, label=y), 100)

    # explain the model's predictions using SHAP values
    explainer = shap.TreeExplainer(model)
    shap_values = explainer.shap_values(X)

    # visualize the first prediction's explaination
    shap.force_plot(explainer.expected_value, shap_values[0, :], X.iloc[0, :])`

In my case, I run XGBoost in a distributed way. After the training is done, we got the model binary. It is hard for us to get the X matrix and put it in the explainer (as in explainer.shap_values(X)). Do you have a plan to export the shap values directly from the xgboost (such that after the distributed training is done, we can easily extract the shap values)?

Thank you
Tien

**Comment by slundberg** [comment]
I am not sure I quite understand what you are asking. If you are looking for the explanations of individual predictions then there is no way around using the feature data for the predictions you want to explain. If you are just looking for the global feature importances (like the sum of the absolute SHAP values). Then you could just use a small subset of X (like 10k examples) and that would give you a great estimate.

We could trace through the model's trees and pull out each leaf as a sample and then put them all back together to get an estimate of the global importance, but I don't have plans to do that right now. (This would essentially allow you to get global feature importance without a dataset)

**Comment by tien-nguyen** [comment]
Hi Scott,

Thank you for your prompt reply! I'm sorry my question was not clear. Let me try to rephrase my question.

My data is about 1.4 TB in LibSVM format. Due to the size of the data, I need to train with distributed Xgboost.

From the example that I see, in order to use explainer, I need to convert my data to a 2D array (n x m: where n is the number of examples, and m is the number of features) - as in explainer.shap_values(X). With the current size of data, I think it's hard to convert to X. Thus my question is if we have another way to get shapes values for each feature without converting my data to X. Essentially, I was looking for the global feature importance of the forests without converting data to 2D dimension.

Based on your question, we don't have a solution for what I'm looking for. Perhaps sampling, as you said, 10 K examples is a way to get the approximated global feature importance.

Thank you.

**Comment by slundberg** [comment]
Makes sense now. Yeah, what you are asking for could be written, but would take a fair amount of effort. I suspect using a subset of the samples will get most of what you want.

**Comment by lidadreamer** [comment]
im also interested in if there will be support for libsvm format dataset
thank u
